library(Seurat)
library(tidyverse)
library(patchwork)
library(reshape2)
library(cowplot) #for plot_grid
library(Matrix) #to do counts on the sparce matrix (not super important)

# script used for all analysis of plate-based scRNAseq data (Figures 1 and Supp Figure 1)
# scRNAseq was performed bt NYGC in replicate

# script requires digital expression matrix (DGE) generated by the dropseq alignment pipeline
# and sample manifest with the metadata for each plate

#setwd("scRNAseq/seurat/NYGC_twoplates")

read_DGE<-function (file_path){
  n<-ncol(read.table(file = file_path))
  prefix <- str_extract(file_path, '(?<=DGE/).*(?=Pool)')
  Q<-read.table(file = file_path, header = TRUE, row.names = 1, colClasses =c("character", rep("numeric", n-1)))
  Q.tb<-as_tibble(rownames_to_column(Q, var = "name"))
  Q.tb<-Q.tb %>% setNames(paste0(prefix,colnames(Q.tb))) %>% rename(name = paste(prefix,"name",sep=""))
  return(Q.tb)
}

join_DGE_Quadrants<-function(Q1,Q2,Q3,Q4) {
  Q1_Q2<-full_join(Q1, Q2, by = "name")
  Q3_Q4<-full_join(Q3, Q4, by = "name")
  merged_quads<-full_join(Q1_Q2, Q3_Q4, by = "name") %>% replace(is.na(.), 0)
  return(merged_quads)
}

####################
# LOAD DATA FROM DGE
####################

P103.Q1<-read_DGE("DGE/Plate103-Q1-Pool.final.q255.dge.txt")
P103.Q2<-read_DGE("DGE/Plate103-Q2-Pool.final.q255.dge.txt")
P103.Q3<-read_DGE("DGE/Plate103-Q3-Pool.final.q255.dge.txt")
P103.Q4<-read_DGE("DGE/Plate103-Q4-Pool.final.q255.dge.txt")

P230.Q1<-read_DGE("DGE/Plate230-Q1-Pool.final.q255.dge.txt")
P230.Q2<-read_DGE("DGE/Plate230-Q2-Pool.final.q255.dge.txt")
P230.Q3<-read_DGE("DGE/Plate230-Q3-Pool.final.q255.dge.txt")
P230.Q4<-read_DGE("DGE/Plate230-Q4-Pool.final.q255.dge.txt")

Plate103.tb<-join_DGE_Quadrants(P103.Q1,P103.Q2,P103.Q3,P103.Q4)
Plate230.tb<-join_DGE_Quadrants(P230.Q1,P230.Q2,P230.Q3,P230.Q4)

###############################################################################
# MERGED SEURAT OBJECTS CREATED FROM EACH PLATE (RETAINING PLATE INFORMATION) #
#############################################################################################################
# FILTER OUT ALL GENES DETECTED IN LESS THAN 3 CELLS FOR CLUSTERING--CHANGE FOR OR GENE EXPRESSION ANALYSIS #
#############################################################################################################
#Create Seurat Objects
scRNA1 <- CreateSeuratObject(counts=data.frame(Plate103.tb, row.names = 1), min.cells=3, min.features=0, project="rep1")
scRNA2 <- CreateSeuratObject(counts=data.frame(Plate230.tb, row.names = 1), min.cells=3, min.features=0, project="rep2")
scRNA.both<-merge(scRNA1, scRNA2, project = "all")

#add metadata
Plate103.meta<-read_delim(file="sample_manifest/plate_00103_single_cell_barcodes.fixed.txt", delim  ="\t", col_names = TRUE)
Plate103.meta<-Plate103.meta %>% select('barcode'='Cell Barcode','cell_name'="Cell Information", "pool_name"='Pool Name') %>% 
  filter(cell_name !='NYGC_Control') %>%
  separate(pool_name,into=c("v1","v2","v3","quad","v5"),sep = "_") %>% 
  separate(cell_name, into=c("zone","v6"),sep="-",remove=FALSE) %>% 
  separate(v6, into=c("sort_type","v7"),sep = "(?<=[A-Za-z])(?=[0-9])") %>% 
  mutate("cell_id1"= paste("Plate103.",quad,sep="")) %>% 
  unite("cell_id",c(cell_id1,barcode), sep=".") %>% select(cell_id,cell_name,zone, sort_type) 

Plate230.meta<-read_delim(file="sample_manifest/plate_00230_sample_manifest.txt", delim  ="\t", col_names = TRUE)
Plate230.meta<-Plate230.meta  %>%rename("cell_name"='Client Sample ID', "sort_type" = 'Cell Lineage', "tissue"='Tissue Type', "quad" = 'Quandrant', "barcode" = "Cell Barcode") %>% 
  filter(cell_name !='NYGC_Control') %>%  mutate("zone"= if_else( tissue == "Zone5","Z5","Z1")) %>% 
  mutate("cell_id1"= paste("Plate230.",quad,sep="")) %>% 
  unite("cell_id",c(cell_id1, barcode), sep=".") %>% 
  select(cell_id,cell_name,zone,sort_type) 


both.meta<-data.frame(bind_rows(Plate103.meta, Plate230.meta)) %>% column_to_rownames('cell_id')
scRNA.both <- AddMetaData(object = scRNA.both, metadata = both.meta, col.name = c("cell_name","zone","sort_type"))

#add mito metadata
mito_meta_df<-read.delim("scRNAseq/seurat/NYGC_twoplates/mito/Plate103_230_mito_meta.tsv", header=TRUE)
scRNA.both <- AddMetaData(object = scRNA.both, metadata = mito_meta_df, col.name = "mito")

#do some QC on the cells
counts_per_cell <- Matrix::colSums(scRNA.both)
counts_per_gene <- Matrix::rowSums(scRNA.both)
genes_per_cell <-Matrix::colSums(scRNA.both@assays$RNA@counts>0)
hist(log10(counts_per_cell+1),main='counts per cell',col='wheat')
hist(log10(genes_per_cell+1), main='genes per cell', col='wheat')
plot(counts_per_cell, genes_per_cell, log='xy', col='wheat')
plot(sort(genes_per_cell), xlab='cell', log='y', main='genes per cell (ordered)')

#cell cycle analysis
# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.  We can
# segregate this list into markers of G2/M phase and markers of S phase
s.genes2<-c("Mcm5","Pcna","Tyms","Fen1","Mcm2","Mcm4","Rrm1","Ung","Gins2","Mcm6","Cdca7","Dtl","Prim1","Uhrf1",
            "Mlf1ip","Hells","Rfc2","Rpa2","Nasp","Rad51ap1","Gmnn","Wdr76","Slbp","Ccne2","Ubr7","Pold3","Msh2","Atad2",
            "Rad51","Rrm2","Cdc45","Cdc6","Exo1","Tipin","Dscc1","Blm","Casp8ap2","Usp1","Clspn","Pola1","Chaf1b","Brip1","E2f8")
g2m.genes2<-c("Hmgb2","Cdk1","Nusap1","Ube2c","Birc5","Tpx2","Top2a","Ndc80","Cks2","Nuf2","Cks1b","Mki67","Tmpo","Cenpf","Tacc3","Fam64a",
              "Smc4","Ccnb2","Ckap2l","Ckap2","Aurkb","Bub1","Kif11","Anp32e","Tubb4b","Gtse1","Kif20b","Hjurp","Cdca3","Hn1","Cdc20","Ttk",
              "Cdc25c","Kif2c","Rangap1","Ncapd2","Dlgap5","Cdca2","Cdca8","Ect2","Kif23","Hmmr","Aurka","Psrc1","Anln","Lbr","Ckap5","Cenpe","Ctcf","Nek2","G2e3","Gas2l3","Cbx5","Cenpa")
median(scRNA.both@meta.data$nCount_RNA)
median(scRNA.both@meta.data$nFeature_RNA)

#Normalize data
scRNA.both %<>% 
  subset(nCount_RNA > 20000 & nFeature_RNA > 1000  & mito < 5) %>%
  NormalizeData(normalization.method = "LogNormalize", scale.factor = 10000) %>%
  FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>% 
  #ScaleData(vars.to.regress = c("S.Score", "G2M.Score")) %>% ##to remove cell cycle genes run 
  ScaleData(features = rownames(scRNA.both)) %>% 
  RunPCA()

#Visually one elbow at 7 another at 10 another at 12 or 13. So 12-13 PCs maximizes the data (but my goal is to find comperable clusters between nfi ko and dev plates, so maybe fewer is better)
DimHeatmap(scRNA.both, dims = 1:5, cells = 100)
JackStraw(scRNA.both, num.replicate = 100) %>% ScoreJackStraw(dims = 1:20) %>% JackStrawPlot(dims = 1:15)
ElbowPlot(scRNA.both)

#Cluster Cells (TSNE and UMAP)
scRNA.both %<>%
  FindNeighbors(dims = 1:11) %>%
  FindClusters(resolution = 0.5) %>%
  RunTSNE(dims = 1:11) %>%
  RunUMAP(dims = 1:11)

#cell cycle
scRNA.both<-CellCycleScoring(scRNA.both,s.features = s.genes2, g2m.features = g2m.genes2, set.ident = FALSE)

# plots of clusters
p.tsne<-TSNEPlot(scRNA.both,  label = T)+ggtitle("Seurat Cluster Identity")+theme(plot.title = element_text(hjust = 0.5))
p.tsne
p.tsne.plate<-TSNEPlot(scRNA.both,group.by = 'orig.ident',label = T, repel = TRUE)+ ggtitle("Replicate") +theme(plot.title = element_text(hjust = 0.5))
p.tsne.plate
p.tsne.sort<-TSNEPlot(scRNA.both, group.by = 'sort_type',label = T, repel = TRUE)+ ggtitle("Sorted Population") +theme(plot.title = element_text(hjust = 0.5))
p.tsne.sort
p.tsne.phase<-TSNEPlot(scRNA.both, group.by = 'Phase',label = T, repel = TRUE)+ ggtitle("Cell Cycle") +theme(plot.title = element_text(hjust = 0.5))
p.tsne.phase

p.umap<-DimPlot(scRNA.both, reduction = "umap", label = T)+ggtitle("Seurat Cluster Identity")+theme(plot.title = element_text(hjust = 0.5))
p.umap
p.umap.plate<-DimPlot(scRNA.both, reduction = "umap",group.by = 'orig.ident',label = T, repel = TRUE)+ ggtitle("Replicate") +theme(plot.title = element_text(hjust = 0.5))
p.umap.plate
p.umap.sort<-DimPlot(scRNA.both, reduction = "umap",group.by = 'sort_type',label = T, repel = TRUE)+ ggtitle("Sorted Population") +theme(plot.title = element_text(hjust = 0.5))
p.umap.sort
p.umap.phase<-DimPlot(scRNA.both, reduction = "umap",group.by = 'Phase',label = T, repel = TRUE)+ ggtitle("Cell Cycle") +theme(plot.title = element_text(hjust = 0.5))
p.umap.phase

grid<-plot_grid(p.umap, p.umap.plate, p.umap.sort, p.umap.phase, nrow = 2,ncol=2)
grid

grid<-plot_grid(p.tsne, p.tsne.plate, p.tsne.sort, p.tsne.phase, nrow = 2,ncol=2)
grid

#NOT ACCURATE BECAUSE 3 CELL CUTOFF!!!!!!!! JUST USED TO IDENTIFY MATURE OSN CELL TYPE
scRNA.both[["percent.olfr"]] <- PercentageFeatureSet(scRNA.both, pattern = "^Olfr")
#saveRDS(scRNA.merged, file = "Plate103_203.merged.rds")

#Identify Clusters
p1<-FeaturePlot(scRNA.both, reduction= "tsne",features = c('Ascl1'), label = T)
p1
p2<-FeaturePlot(scRNA.both, reduction= "tsne",features = c('Sox2'), label = T)
p2
p3<-FeaturePlot(scRNA.both, reduction= "tsne",features = c('Neurog1'), label = T)
p3
p4<-FeaturePlot(scRNA.both, reduction= "tsne",features = c('Neurod1'), label = T)
p4
p5<-FeaturePlot(scRNA.both, reduction= "tsne",features = c('Tex15'), label = T)
p5
p6<-FeaturePlot(scRNA.both, reduction= "tsne",features = c('Gng8'), label = T)
p6
p7<-FeaturePlot(scRNA.both, reduction= "tsne",features = c('Gap43'), label = T)
p7
p8<-FeaturePlot(scRNA.both, reduction= "tsne",features = c('Omp'), label = T)
p8
#p8<-FeaturePlot(scRNA.both, reduction= "tsne",features = "percent.olfr",label = T)
#p8
grid<-plot_grid(p1,p2,p3,p4,p5,p6,p7,p8, nrow = 2,ncol=4)
grid

###########################################################
# look at Gng8 expression across cell types (Supp Figure 7D)
###########################################################

tsne.Gng8.z5<-FeaturePlot(subset(scRNA.both, subset = zone == "Z5"), reduction= "tsne",features = c('Gng8'), label = T, max.cutoff = 4.5)+ labs(title = "zone5") & scale_color_viridis_c() 
tsne.Gng8.z1<-FeaturePlot(subset(scRNA.both, subset = zone == "Z1"), reduction= "tsne",features = c('Gng8'), label = T, max.cutoff = 4.5)+ labs(title = "zone1") & scale_color_viridis_c() 
grid<-plot_grid(tsne.Gng8.z1,tsne.Gng8.z5, nrow = 1,ncol=2)
grid
save_plot("Tsne_Gng8_zonal.pdf",grid,base_height = 4,base_width = 8)

tsne.Gng8<-FeaturePlot(scRNA.both, reduction= "tsne",features = c('Gng8'), label = T)+ labs(title = "Gng8") & scale_color_viridis_c() 
tsne.Gng8
save_plot("Tsne_Gng8.pdf",tsne.Gng8,base_height = 4,base_width = 4)

#############################################################
#############################################################

# Define an order of cluster identities. According to developmental trajectory. And set levels
my_levels <- c(3,5,0,1,2,4)
Idents(scRNA.both) <- factor(Idents(scRNA.both), levels= my_levels)

p.violin.olfr<-VlnPlot(scRNA.both, features = "percent.olfr", log = TRUE, pt.size = 0) + xlab("Cluster Identity")
p.violin.olfr

#Label clusters by developmental trajectory (using clusters defined by 3 cell cutoff data)
scRNA.both@meta.data <- 
  scRNA.both@meta.data %>%
  as.data.frame() %>% 
  rownames_to_column(var = 'cell_id') %>%
  mutate('to_join' = as.numeric(as.character(seurat_clusters))) %>%
  left_join(data.frame('to_join' = c(0,1,2,3,4,5),
                       'cell_type' = c('INP2','INP3','iOSN','GBC','OSN','INP1')),
            by = 'to_join') %>% 
  dplyr::select(-to_join) %>%
  column_to_rownames(var = 'cell_id') 

#how many cell in each cell-type,zone
scRNA.both@meta.data %>% group_by(cell_type,zone)%>%  summarise(n())
#cell_type zone  `n()`
#<fct>     <chr> <int>
#1 GBC       Z1       48
#2 GBC       Z5       43
#3 INP1      Z1       16
#4 INP1      Z5       20
#5 INP2      Z1       59
#6 INP2      Z5       87
#7 INP3      Z1       64
#8 INP3      Z5       62
#9 iOSN      Z1       71
#10 iOSN      Z5       37
#11 OSN       Z1       49
#12 OSN       Z5       35



#look at differentially expressed genes
cluster1.markers <- FindMarkers(scRNA.both, ident.1 = 1, min.pct = 0.25) %>% rownames_to_column(var="gene") %>% filter(p_val_adj < 0.05) %>% mutate("cluster_name"="INP3")
head(cluster1.markers, n = 10)
cluster2.markers <- FindMarkers(scRNA.both, ident.1 = 2, min.pct = 0.25) %>% rownames_to_column(var="gene") %>% filter(p_val_adj < 0.05) %>% mutate("cluster_name"="iOSN")
head(cluster2.markers, n = 10)
cluster0.markers <- FindMarkers(scRNA.both, ident.1 = 0, min.pct = 0.25) %>% rownames_to_column(var="gene") %>% filter(p_val_adj < 0.05) %>% mutate("cluster_name"="INP2")
head(cluster0.markers, n = 10)
cluster3.markers <- FindMarkers(scRNA.both, ident.1 = 3, min.pct = 0.25) %>% rownames_to_column(var="gene") %>% filter(p_val_adj < 0.05) %>% mutate("cluster_name"="GBC")
head(cluster3.markers, n = 10)
cluster4.markers <- FindMarkers(scRNA.both, ident.1 = 4, min.pct = 0.25) %>% rownames_to_column(var="gene") %>% filter(p_val_adj < 0.05) %>% mutate("cluster_name"="OSN")
head(cluster4.markers, n = 10)
cluster5.markers <- FindMarkers(scRNA.both, ident.1 = 5, min.pct = 0.25) %>% rownames_to_column(var="gene") %>% filter(p_val_adj < 0.05) %>% mutate("cluster_name"="INP1")
head(cluster5.markers, n = 10)
cluster_markers_dim11_mito6<-bind_rows(cluster4.markers,cluster2.markers,cluster1.markers,cluster0.markers,cluster5.markers,cluster3.markers)


# find markers for every cluster compared to all remaining cells, report only the positive ones
scRNA.both.markers <- FindAllMarkers(scRNA.both, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
scRNA.both.markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_logFC) %>% View()

####make a heatmap of selected olfactory lineage genes (Supp Figure 1C)
olfactory_genes<-c("Kit","Ascl1","Sox2","Pax6","Neurog1","Neurod1","Lhx2","Sox11","Ldb1","Zfp423","Gng8","Ebf2","Ebf3","Ebf1","Ncam1","Atf5","Gap43","Ebf4","Adcy3","Cnga2","Gng13","Gnal","Omp")
data <- scRNA.both
Idents(data) <- data@meta.data$cell_type

plot<-DoHeatmap(data, features = olfactory_genes,angle = 0,hjust = (0.5)) + 
  scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n = 10, name = "RdBu")) )+
  theme(axis.text.y = element_text(color = "black",size = 11,family = "Palatino"), text = element_text(size = 12,family = "Palatino", color="black"))
plot
#save_plot("Marker_heatmap.pdf",plot,base_height = 10,base_width = 12)

Idents(data) <- paste(data@meta.data$cell_type, data@meta.data$zone, sep = "_")
my_levels <- c("GBC_Z5","INP1_Z5","INP2_Z5","INP3_Z5","iOSN_Z5","OSN_Z5","GBC_Z1","INP1_Z1","INP2_Z1","INP3_Z1","iOSN_Z1","OSN_Z1")
Idents(data) <- factor(Idents(data), levels= my_levels)

Nfis<-c("Nfia","Nfib","Nfix")
plot2<-DoHeatmap(data, features = Nfis,angle = 90,hjust = (0)) + 
  scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n = 10, name = "RdBu")) )+
  theme(axis.text.y = element_text(color = "black",size = 11,family = "Palatino"), text = element_text(size = 12,family = "Palatino", color="black"))
plot2
#save_plot("Nfi_heatmap.pdf",plot2,base_height = 10,base_width = 12)


## perform default DE analysis non-parameteric Wilcoxon rank sum test.
scRNA.both.var<-scRNA.both
Idents(scRNA.both.var) <- paste(Idents(scRNA.both.var), scRNA.both.var@meta.data$zone, sep = "_")
zonal_clust4 <- FindMarkers(scRNA.both.var, ident.1 = "4_Z5", ident.2 = "4_Z1") %>% rownames_to_column(var="gene") %>% filter(p_val_adj < 0.05) %>% mutate("cluster_name"="OSN")
zonal_clust4
zonal_clust2 <- FindMarkers(scRNA.both.var, ident.1 = "2_Z5", ident.2 = "2_Z1") %>% rownames_to_column(var="gene") %>% filter(p_val_adj < 0.05) %>% mutate("cluster_name"="iOSN")
zonal_clust2
zonal_clust1 <- FindMarkers(scRNA.both.var, ident.1 = "1_Z5", ident.2 = "1_Z1") %>% rownames_to_column(var="gene") %>% filter(p_val_adj < 0.05) %>% mutate("cluster_name"="INP3")
zonal_clust1
zonal_clust0 <- FindMarkers(scRNA.both.var, ident.1 = "0_Z5", ident.2 = "0_Z1") %>% rownames_to_column(var="gene") %>% filter(p_val_adj < 0.05) %>% mutate("cluster_name"="INP2")
zonal_clust0
zonal_clust5 <- FindMarkers(scRNA.both.var, ident.1 = "5_Z5", ident.2 = "5_Z1") %>% rownames_to_column(var="gene") %>% filter(p_val_adj < 0.05) %>% mutate("cluster_name"="INP1")
zonal_clust5
zonal_clust3 <- FindMarkers(scRNA.both.var, ident.1 = "3_Z5", ident.2 = "3_Z1") %>% rownames_to_column(var="gene") %>% filter(p_val_adj < 0.05) %>% mutate("cluster_name"="GBC")
zonal_clust3
cluster_diff_zonal_sig_alpha0.05<-bind_rows(zonal_clust4,zonal_clust2,zonal_clust1,zonal_clust0,zonal_clust5,zonal_clust3)
#write_delim(cluster_diff_zonal_sig_alpha0.05,"DE_Z5vZ1_bycluster_alpha0.05_dim11_mito5.tsv", delim = "\t")

######################################################################
# REDO SUBSETTING, NORM WITHOUT 3 CELL CUTOFF TO GET ACCURATE OR COUNT
######################################################################

#Create Seurat Objects
scRNA1.OR <- CreateSeuratObject(counts=data.frame(Plate103.tb, row.names = 1), min.cells=0, min.features=0, project="rep1")
scRNA2.OR <- CreateSeuratObject(counts=data.frame(Plate230.tb, row.names = 1), min.cells=0, min.features=0, project="rep2")
scRNA.both.OR<-merge(scRNA1.OR, scRNA2.OR, project = "all")
#add cell type metadata
scRNA.both.OR <- AddMetaData(object = scRNA.both.OR, metadata = both.meta, col.name = c("cell_name","zone","sort_type"))
#add mito metadata
scRNA.both.OR <- AddMetaData(object = scRNA.both.OR, metadata = mito_meta_df, col.name = "mito")
#Normalize data
scRNA.both.OR %<>% 
  subset(nCount_RNA > 20000 & nFeature_RNA > 1000  & mito < 5) %>%
  NormalizeData(normalization.method = "LogNormalize", scale.factor = 10000) %>%
  FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>% 
  #ScaleData(vars.to.regress = c("S.Score", "G2M.Score")) %>% ##to remove cell cycle genes run 
  ScaleData() %>% 
  RunPCA()
#Cluster Cells (TSNE and UMAP)
scRNA.both.OR %<>%
  FindNeighbors(dims = 1:11) %>%
  FindClusters(resolution = 0.5) %>%
  RunTSNE(dims = 1:11) %>%
  RunUMAP(dims = 1:11)
#cell cycle
scRNA.both.OR<-CellCycleScoring(scRNA.both.OR,s.features = s.genes2, g2m.features = g2m.genes2, set.ident = FALSE)

# plots of clusters
p.tsne<-TSNEPlot(scRNA.both.OR,  label = T)+ggtitle("Seurat Cluster Identity")+theme(plot.title = element_text(hjust = 0.5))
p.tsne
p.tsne.plate<-TSNEPlot(scRNA.both.OR,group.by = 'orig.ident',label = T, repel = TRUE)+ ggtitle("Replicate") +theme(plot.title = element_text(hjust = 0.5))
p.tsne.plate
p.tsne.sort<-TSNEPlot(scRNA.both.OR, group.by = 'sort_type',label = T, repel = TRUE)+ ggtitle("Sorted Population") +theme(plot.title = element_text(hjust = 0.5))
p.tsne.sort
p.tsne.phase<-TSNEPlot(scRNA.both.OR, group.by = 'Phase',label = T, repel = TRUE)+ ggtitle("Cell Cycle") +theme(plot.title = element_text(hjust = 0.5))
p.tsne.phase

scRNA.both.OR[["percent.olfr"]] <- PercentageFeatureSet(scRNA.both.OR, pattern = "^Olfr")
# Define an order of cluster identities. According to developmental trajectory. And set levels
my_levels <- c(3,5,0,1,2,4)
Idents(scRNA.both.OR) <- factor(Idents(scRNA.both.OR), levels= my_levels)

p.violin.olfr<-VlnPlot(scRNA.both.OR, features = "percent.olfr", log = TRUE, pt.size = 0) + xlab("Cluster Identity")
p.violin.olfr

#########################################
### Assess OR Total Expression Levels ###
#########################################

# zonal OR color palettes
zonal_fine_colors<-c("#FF0000","#FFCC00","#33CC33","#00CCCC","#0000FF")
zonal_fine_colors_fish<-c("#999999","#FF0000","#FFCC00","#33CC33","#00CCCC","#0000FF")
zonal_fine_colors_fish_inv<-c("#0000FF","#00CCCC","#33CC33","#FFCC00","#FF0000","#999999")
zonal_fine_colors_fish_new3_inv<-c('#0505B7','#077FE2','#23BD7D','#FFCC00','#EC001D','#860000')
zonal_fine_colors_fish_new3_inv_white<-c('#0505B7','#077FE2','#23BD7D','#FFCC00','#EC001D','#860000',"white")
zonal_fine_colors_fish_new3_inv_no5<-c('#077FE2','#23BD7D','#FFCC00','#EC001D','#860000')
zonal_fine_colors_fish_new3_inv_no5_white<-c('#077FE2','#23BD7D','#FFCC00','#EC001D','#860000', "white")

#extract normalised data and raw counts (to filter on UMI's < 3)
norm.data<-scRNA.both.OR@assays$RNA@data[str_detect(rownames(scRNA.both.OR), 'Olfr'),] %>% 
  as.data.frame() %>% rownames_to_column(var = 'gene_name') %>% 
  melt(id.vars = 'gene_name', variable.name = 'cell_id') %>% rename("norm_count" = value)
#Olfr_genes<-unique(norm.data$gene_name)
#Olfr_genes2<-rownames(scRNA.both@assays$RNA@counts[grep('Olfr',rownames(scRNA.both@assays$RNA@counts)),])
count.data<-scRNA.both.OR@assays$RNA@counts[str_detect(rownames(scRNA.both.OR), 'Olfr'),] %>% 
  as.data.frame() %>% rownames_to_column(var = 'gene_name') %>% 
  melt(id.vars = 'gene_name', variable.name = 'cell_id') %>% rename("count" = value)

#count.data<-scRNA.both@assays$RNA@counts[rownames(scRNA.both@assays$RNA@counts) %in% Olfr_genes2,] %>% 
#  as.data.frame() %>% rownames_to_column(var = 'gene_name') %>% 
#  melt(id.vars = 'gene_name', variable.name = 'cell_id') %>% rename("count" = value)


scRNA.bothOlfr<-full_join(norm.data,count.data) %>% 
  left_join(rownames_to_column(scRNA.both@meta.data, var = 'cell_id'), by='cell_id')
scRNA.bothOlfr<-scRNA.bothOlfr %>%
  mutate("cutoff_3umi"=if_else(count <3, 0, norm_count)) %>%
  mutate("cutoff_5umi"=if_else(count <5, 0, norm_count))

###########################
###########################
# count OR co-expression

scRNA.bothOlfr_coexpression<-scRNA.bothOlfr %>% dplyr::select(gene_name,cell_id,zone,cell_type,cutoff_3umi) %>% 
  drop_na() %>% filter(cutoff_3umi>0) %>% 
  group_by(cell_id,zone,cell_type) %>% filter(cell_type %in% c("INP3","iOSN","OSN"))%>%
  summarise("n_ORs"=n()) %>% mutate("all"="all")

scRNA.bothOlfr_coexpression$cell_type<-factor(scRNA.bothOlfr_coexpression$cell_type, levels=c("INP3","iOSN","OSN"))

p.n_OR<-ggplot(scRNA.bothOlfr_coexpression,aes(x=all,y=n_ORs, fill=zone, color=zone))+ 
  geom_boxplot(alpha = 0.7, outlier.shape = NA, fatten=4)+
  geom_point(position=position_jitterdodge(jitter.height=0.2,jitter.width = 0.2 ),alpha=0.7)+xlab("cluster identity")+
  labs(fill="dissected zone",color="dissected zone")+ 
  ggtitle("ORs expressed per cell (3umi cutoff)")+ 
  scale_fill_manual(values=c("#FF3300","#3399FF")) + 
  scale_color_manual(values=c("#990000","#003399")) + ylim(c(0,22))+
  facet_grid(cols=vars(cell_type))
p.n_OR
ggsave(filename = "boxplot.n_OR_coexpressed.pdf",p.n_OR, height = 5, width = 5, units = "in", useDingbats=FALSE)



#average
scRNA.bothOlfr_coexpression_stats<-scRNA.bothOlfr_coexpression %>% group_by(zone,cell_type) %>% 
  summarise("avg_n_OR"=mean(n_ORs),"median_n_OR"=median(n_ORs)) 

stat.tb_1<-filter(scRNA.bothOlfr_coexpression, cell_type == "INP3" & zone =="Z1") 
stat.tb_5<-filter(scRNA.bothOlfr_coexpression, cell_type == "INP3" & zone =="Z5") 

wilcox.test(stat.tb_1$n_ORs, stat.tb_5$n_ORs)
#Wilcoxon rank sum test with continuity correction
#W = 648.5, p-value = 0.002666

stat.tb_1<-filter(scRNA.bothOlfr_coexpression, cell_type == "iOSN" & zone =="Z1") 
stat.tb_5<-filter(scRNA.bothOlfr_coexpression, cell_type == "iOSN" & zone =="Z5") 

wilcox.test(stat.tb_1$n_ORs, stat.tb_5$n_ORs)
#Wilcoxon rank sum test with continuity correction
#W = 808, p-value = 0.1241

stat.tb_1<-filter(scRNA.bothOlfr_coexpression, cell_type == "OSN" & zone =="Z1") 
stat.tb_5<-filter(scRNA.bothOlfr_coexpression, cell_type == "OSN" & zone =="Z5") 
wilcox.test(stat.tb_1$n_ORs, stat.tb_5$n_ORs)
#Wilcoxon rank sum test with continuity correction
#W = 688, p-value = 0.1956


####fract co-expression
co_express.tb<-scRNA.bothOlfr_coexpression %>% 
  mutate("coexpress"=if_else(n_ORs >1,"yes","no")) %>% group_by(zone,cell_type) %>%
  summarise("fract_coexpression"= sum(coexpress == "yes")/n())
co_express.tb

p.fract_OR<-ggplot(co_express.tb %>% mutate("all"="all"),aes(x=all,y=fract_coexpression, color=zone, fill=zone))+ 
  geom_bar(stat="identity", position = "dodge", alpha=0.7)+ylim(0,1)  +
  labs(fill="dissected zone",color="dissected zone")+ 
  ggtitle("fraction of cells co-expressing ORs (3umi cutoff)")+ 
  scale_fill_manual(values=c("#FF3300","#3399FF")) + 
  scale_color_manual(values=c("#990000","#003399")) +
  facet_grid(cols=vars(cell_type))

p.fract_OR
ggsave(filename = "barplot.fract_OR_coexpressed.pdf",p.fract_OR, height = 5, width = 5, units = "in", useDingbats=FALSE)

############################
####### BOX PLOTS ##########
############################

scRNA.pass3UMI<-scRNA.bothOlfr %>%
  group_by(cell_id) %>% tally(cutoff_3umi) %>% 
  rename("total_norm_count_3UMI"=n) %>% right_join(scRNA.bothOlfr) %>%
  select(cell_id,total_norm_count_3UMI,gene_name,orig.ident,cell_type,Phase,zone,cutoff_3umi) %>% 
  drop_na() %>% distinct() #has all ORs for all cells


scRNA.pass5UMI<-scRNA.bothOlfr %>%
  group_by(cell_id) %>% tally(cutoff_5umi) %>% 
  rename("total_norm_count_5UMI"=n) %>% right_join(scRNA.bothOlfr) %>%
  select(cell_id,total_norm_count_5UMI,gene_name,orig.ident,cell_type,Phase,zone,cutoff_5umi) %>% 
  drop_na() %>% distinct()

###plots with 3UMI cutoff
working<-select(scRNA.pass3UMI,cell_id,cell_type,orig.ident,total_norm_count_3UMI,cutoff_3umi, zone)%>% distinct()
ggplot(working,aes(x=factor(cell_type, levels=c("GBC","INP1","INP2","INP3","iOSN","OSN")),y=total_norm_count_3UMI, fill=zone))+ 
  geom_boxplot()+xlab("cluster identity")+labs(fill="dissected zone")+ ggtitle("Total OR expression/cell")+ scale_fill_manual(values=c("#FF0000","#0033FF"))

working<-select(scRNA.pass3UMI,cell_id,orig.ident,cell_type,total_norm_count_3UMI, zone) %>% filter(total_norm_count_3UMI >0)  %>% distinct()
p3.total<-ggplot(working,aes(x=factor(cell_type, levels=c("GBC","INP1","INP2","INP3","iOSN","OSN")),y=total_norm_count_3UMI,color=orig.ident))+ 
  geom_jitter(size=1,alpha = 0.8,height=0.3)+xlab("cluster identity")+labs(color="plate")+ ggtitle("Total OR expression/cell")
p3.total
p3.total2<-ggplot(working,aes(x=factor(cell_type, levels=c("GBC","INP1","INP2","INP3","iOSN","OSN")),y=total_norm_count_3UMI,color=zone))+ 
  geom_jitter(size=1,alpha = 0.5,height=0.3)+xlab("cluster identity")+labs(color="plate")+ ggtitle("Total OR expression/cell")+ 
  scale_color_manual(values=c("#FF0000","#0033FF")) + ylab("total norm count OR")
p3.total2
p3.total3<-ggplot(working,aes(x=factor(cell_type, levels=c("GBC","INP1","INP2","INP3","iOSN","OSN")),y=total_norm_count_3UMI))+ 
  geom_boxplot(size=1,alpha = 0.5)+xlab("cluster identity")+labs(color="plate")+ ggtitle("Total OR expression/cell")+ 
  scale_color_manual(values=c("#FF0000","#0033FF")) + ylab("total norm count OR")
p3.total3



working<-select(scRNA.pass3UMI,cell_id,orig.ident,cell_type,total_norm_count_3UMI,cutoff_3umi,zone) %>% filter(cutoff_3umi >0)  %>% distinct()
p3.indv<-ggplot(working,aes(x=factor(cell_type, levels=c("GBC","INP1","INP2","INP3","iOSN","OSN")),y=cutoff_3umi, color=orig.ident))+ 
  geom_jitter(size=1,alpha = 0.8,height=0.1)+xlab("cluster identity")+labs(color="plate")+ ggtitle("Individual OR expression")
p3.indv
p3.indv2<-ggplot(working,aes(x=factor(cell_type, levels=c("GBC","INP1","INP2","INP3","iOSN","OSN")),y=cutoff_3umi, color=zone))+ 
  geom_jitter(size=1,alpha = 0.5,height=0.1)+xlab("cluster identity")+labs(color="plate")+ ggtitle("Individual OR expression")+ 
  scale_color_manual(values=c("#FF0000","#0033FF"))+ ylab("norm count individual OR")
p3.indv2
p3.indv3<-ggplot(working,aes(x=factor(cell_type, levels=c("GBC","INP1","INP2","INP3","iOSN","OSN")),y=cutoff_3umi))+ 
  geom_boxplot(size=1,alpha = 0.5,height=0.1)+xlab("cluster identity")+labs(color="plate")+ ggtitle("Individual OR expression")+ 
  scale_color_manual(values=c("#FF0000","#0033FF"))+ ylab("norm count individual OR")
p3.indv3


grid<-plot_grid(p3.total2,p3.indv2, ncol=2)
grid
#save_plot("Boxplots_3UMI_bothPlates_dim10.pdf",grid,base_height = 8,base_width = 12)

###plots with 5UMI cutoff
working<-select(scRNA.pass5UMI,cell_id,cell_type,orig.ident,total_norm_count_5UMI,cutoff_5umi, zone)%>% distinct()
ggplot(working,aes(x=factor(cell_type, levels=c("GBC","INP1","INP2","INP3","iOSN","OSN")),y=total_norm_count_5UMI, fill=zone))+ 
  geom_boxplot()+xlab("cluster identity")+labs(fill="dissected zone")+ ggtitle("Total OR expression/cell")+ scale_fill_manual(values=c("#FF0000","#0033FF"))

working<-select(scRNA.pass5UMI,cell_id,orig.ident,cell_type,total_norm_count_5UMI, zone) %>% filter(total_norm_count_5UMI >0)  %>% distinct()
p3.total<-ggplot(working,aes(x=factor(cell_type, levels=c("GBC","INP1","INP2","INP3","iOSN","OSN")),y=total_norm_count_5UMI,color=orig.ident))+ 
  geom_jitter(size=1,alpha = 0.8,height=0.3)+xlab("cluster identity")+labs(color="plate")+ ggtitle("Total OR expression/cell")
p3.total
p3.total2<-ggplot(working,aes(x=factor(cell_type, levels=c("GBC","INP1","INP2","INP3","iOSN","OSN")),y=total_norm_count_5UMI,color=zone))+ 
  geom_jitter(size=1,alpha = 0.5,height=0.3)+xlab("cluster identity")+labs(color="plate")+ ggtitle("Total OR expression/cell")+ 
  scale_color_manual(values=c("#FF0000","#0033FF")) + ylab("total norm count OR")
p3.total2

working<-select(scRNA.pass5UMI,cell_id,orig.ident,cell_type,total_norm_count_5UMI,cutoff_5umi,zone) %>% filter(cutoff_5umi >0)  %>% distinct()
p3.indv<-ggplot(working,aes(x=factor(cell_type, levels=c("GBC","INP1","INP2","INP3","iOSN","OSN")),y=cutoff_5umi, color=orig.ident))+ 
  geom_jitter(size=1,alpha = 0.8,height=0.1)+xlab("cluster identity")+labs(color="plate")+ ggtitle("Individual OR expression")
p3.indv
p3.indv2<-ggplot(working,aes(x=factor(cell_type, levels=c("GBC","INP1","INP2","INP3","iOSN","OSN")),y=cutoff_5umi, color=zone))+ 
  geom_jitter(size=1,alpha = 0.5,height=0.1)+xlab("cluster identity")+labs(color="plate")+ ggtitle("Individual OR expression")+ 
  scale_color_manual(values=c("#FF0000","#0033FF"))+ ylab("norm count individual OR")
p3.indv2

grid<-plot_grid(p3.total2,p3.indv2, ncol=2)
grid
#save_plot("Boxplots_5UMI_bothPlates_dim10.pdf",grid,base_height = 8,base_width = 12)

###################################
## COMBINED BARPLOTS BY CELL TYPE #
###################################
#make data frame with zonal annotation and all ORs over 3UMI
#cells not expressing ORs get one line with NA's in gene_name and ORzone
#to filter out cells not expressing ORs run drop_na()

scRNA.pass3UMI.zone<-read.delim(file="zonal_paper/annotation/zonal_anno_2020_fine.tsv", header =FALSE, sep="\t") %>%
  rename("gene_name"=V1,"ORzone"=V6) %>% select(gene_name,ORzone) %>% filter(!(gene_name == "Olfr190" & ORzone == 5)) %>% right_join(scRNA.pass3UMI, by = 'gene_name')
all.cells.temp<-select(scRNA.pass3UMI,cell_id,total_norm_count_3UMI,orig.ident,cell_type,Phase,zone) %>% distinct()
scRNA.pass3UMI.zone<-scRNA.pass3UMI.zone %>% 
  filter(cutoff_3umi >0) %>% full_join(all.cells.temp) 

#calculate total count per zone for each cell type
tb.total.zone<-scRNA.pass3UMI.zone %>% 
  group_by(cell_type,zone,ORzone) %>% summarize(total_cutoff_3umi=sum(cutoff_3umi))%>% complete(ORzone, nesting(cell_type, zone))%>% 
  right_join(tibble(zone=c(rep("Z1",6),rep("Z5",6)),ORzone=c("classI","1","2","3","4","5","classI","1","2","3","4","5"))) %>%
  replace_na(list(total_cutoff_3umi=0)) 

#now correcting for number of cells contributing to count
#adjust to 100 cells/each
tb.total.zone.adj<-scRNA.both@meta.data %>%
  group_by(cell_type,zone) %>% 
  summarise(cells=n()) %>% mutate(corr_fact = 100/cells) %>%
  right_join(tb.total.zone) %>% 
  mutate(total_cutoff_3umi_adjusted = total_cutoff_3umi*corr_fact)

barplot_facet<-ggplot(data = tb.total.zone.adj, aes(x = zone, y = total_cutoff_3umi_adjusted , fill=factor(ORzone ,levels=c("5","4","3","2","1","classI") ))) +
  geom_bar(position = 'stack', stat="identity", color="gray2",size=0.2) + 
  scale_fill_manual(values=zonal_fine_colors_fish_new3_inv)+
  facet_grid(cols = vars(factor(cell_type, levels =c("GBC","INP1","INP2","INP3","iOSN","OSN") )))+
  labs(fill= "OR zone")+ylab("combined OR expression (normalized counts)")+ scale_x_discrete(labels=c("z1", "z4-5"))+
  theme(strip.text.x = element_text(size = 12, colour = "black",family = "Palatino"), axis.title.x = element_blank(),
        axis.text.x = element_text(color="black", size=12,family = "Palatino",angle = 90), axis.text.y = element_text(color="black", size=9,family = "Palatino"),
        axis.title.y = element_text(color="black", size=12,family = "Palatino"),legend.text = element_text(size = 12,family = "Palatino"), legend.title = element_text(size = 14,family = "Palatino"))
barplot_facet
#save_plot("Barplots_combined_WTzone1vzone5_all_cell_types_dim11_mito5_adjusted.pdf",barplot_facet, base_height = 6, base_width = 10)

##########################################################
## COMBINED BARPLOTS BY CELL TYPE not by zone (Figure 1B)#
##########################################################

#make data frame with zonal annotation and all ORs over 3UMI
#cells not expressing ORs get one line with NA's in gene_name and ORzone
#to filter out cells not expressing ORs run drop_na()


all.cells.temp<-select(scRNA.pass3UMI,cell_id,total_norm_count_3UMI,orig.ident,cell_type) %>% distinct()
box<-ggplot(all.cells.temp, aes(cell_type,total_norm_count_3UMI, fill = cell_type, color = cell_type))+ 
  geom_boxplot(outlier.shape = NA, alpha = 0.8)+ylim(0,13)+coord_flip()+ scale_fill_manual(values = c("#9999FF","#CC99FF","#9900CC","#660099","#330066","#000033"))+
  scale_color_manual(values = c("#9999FF","#CC99FF","#9900CC","#660099","#330066","#000033"))
box
save_plot("Barplots_horizontal_cell_type_total_norm_count_3UMI_2.pdf",box, base_height = 6, base_width = 4)


##############################################################
## INDIVIDUAL BARPLOTS OF OR Expression per cell (Figure1C-D)#
##############################################################
#includes unknown ORs
#plot p_zone like this "Z5", p_cell_type like "INP3" and p_title like this "Olfr expression in zone5 INP3 cells", and p_xlab like this: "individual zone5 INP3 cells"
#plot p_palette like this c(zonal_fine_colors_fish_new3_inv, "white"), plot p_labels like this c("zone5", "zone4", "zone3","zone2", "zone1", "classI", "unknown")
plot_individual<-function(p_zone,p_cell_type,p_title,p_xlab,p_palette,p_labels){
  ggplot(data = filter(scRNA.pass3UMI.zone,zone == p_zone & cell_type == p_cell_type ),aes(x = reorder(cell_id, -total_norm_count_3UMI), y = cutoff_3umi , fill=factor(ORzone ,levels=c("5","4","3","2","1","classI") ))) +
    geom_bar(position = 'stack', stat="identity", col= "black", size=0.3) + ylim(c(0,25))+
    theme(axis.title.x=element_text(color="black",size=12,family = "Palatino"),axis.text.x=element_blank(),axis.text.y=element_text(size = 9, color="black",family = "Palatino"), 
          axis.title.y = element_text(size=12,family = "Palatino", color="black"),plot.title = element_text(hjust = 0.5,family = "Palatino",size=14), 
          legend.text=element_text(family = "Palatino", size=12), legend.title=element_text(size=14,family = "Palatino"))+
    ggtitle(p_title)+scale_x_discrete(drop=FALSE)+xlab(p_xlab)+
    scale_fill_manual(values=p_palette, labels = p_labels)+labs(fill= "OR zone") +ylab("normalized OR expression")+theme(legend.position = "none")
}
#function that excludes unknown ORs
#to exclude unknown ORs: data=drop_na(filter(scRNA.pass3UMI.zone,zone == "Z5" & cell_type == "INP3" ) 
#and: scale_fill_manual(values=zonal_fine_colors_fish_new3_inv)
plot_individual_NOunknown<-function(p_zone,p_cell_type,p_title,p_xlab,p_palette,p_labels){
  ggplot(data = drop_na(filter(scRNA.pass3UMI.zone,zone == p_zone & cell_type == p_cell_type )),aes(x = reorder(cell_id, -total_norm_count_3UMI), y = cutoff_3umi , fill=factor(ORzone ,levels=c("5","4","3","2","1","classI") ))) +
    geom_bar(position = 'stack', stat="identity", col= "black", size=0.3) + ylim(c(0,25))+
    theme(axis.title.x=element_text(color="black",size=12,family = "Palatino"),axis.text.x=element_blank(),axis.text.y=element_text(size = 9, color="black",family = "Palatino"), 
          axis.title.y = element_text(size=12,family = "Palatino", color="black"),plot.title = element_text(hjust = 0.5,family = "Palatino",size=14), 
          legend.text=element_text(family = "Palatino", size=12), legend.title=element_text(size=14,family = "Palatino"))+
    ggtitle(p_title)+scale_x_discrete(drop=FALSE)+xlab(p_xlab)+
    scale_fill_manual(values=p_palette, labels = p_labels)+labs(fill= "OR zone") +ylab("normalized OR expression")+theme(legend.position = "none")
}

ggplot(data = filter(scRNA.pass3UMI.zone,zone == "Z1" & cell_type == "OSN" ),aes(x = reorder(cell_id, -total_norm_count_3UMI), y = cutoff_3umi , fill=factor(ORzone ,levels=c("5","4","3","2","1","classI") ))) +
  geom_bar(position = 'stack', stat="identity", col= "black", size=0.3)
#INP3
inp5<-plot_individual("Z5","INP3","Olfr expression in zone5 INP3 cells","individual zone5 INP3 cells",zonal_fine_colors_fish_new3_inv_white,c( "zone5","zone4", "zone3","zone2", "zone1", "classI", "unknown"))
inp5
inp1<-plot_individual("Z1","INP3","Olfr expression in zone1 INP3 cells","individual zone1 INP3 cells",zonal_fine_colors_fish_new3_inv_no5_white,c("zone4", "zone3","zone2", "zone1", "classI", "unknown"))
inp1
inp5_drop<-plot_individual_NOunknown("Z5","INP3","Olfr expression in zone5 INP3 cells","individual zone5 INP3 cells",zonal_fine_colors_fish_new3_inv,c( "zone5","zone4", "zone3","zone2", "zone1", "classI"))
inp5_drop
inp1_drop<-plot_individual_NOunknown("Z1","INP3","Olfr expression in zone1 INP3 cells","individual zone1 INP3 cells",zonal_fine_colors_fish_new3_inv_no5,c("zone4", "zone3","zone2", "zone1", "classI", "unknown"))
inp1_drop
#iOSN
iosn5<-plot_individual("Z5","iOSN","Olfr expression in zone5 iOSN cells","individual zone5 iOSN cells",zonal_fine_colors_fish_new3_inv_white,c("zone5", "zone4", "zone3","zone2", "zone1", "classI", "unknown"))
iosn5
iosn1<-plot_individual("Z1","iOSN","Olfr expression in zone1 iOSN cells","individual zone1 iOSN cells",zonal_fine_colors_fish_new3_inv_white,c("zone5","zone4", "zone3","zone2", "zone1", "classI", "unknown"))
iosn1
iosn5_drop<-plot_individual_NOunknown("Z5","iOSN","Olfr expression in zone5 iOSN cells","individual zone5 iOSN cells",zonal_fine_colors_fish_new3_inv_white,c("zone5", "zone4", "zone3","zone2", "zone1", "classI", "unknown"))
iosn1_drop<-plot_individual_NOunknown("Z1","iOSN","Olfr expression in zone1 iOSN cells","individual zone1 iOSN cells",zonal_fine_colors_fish_new3_inv_white,c("zone5","zone4", "zone3","zone2", "zone1", "classI", "unknown"))
#OSN
osn5<-plot_individual("Z5","OSN","Olfr expression in zone5 OSN cells","individual zone5 OSN cells",c('#0505B7','#077FE2','#23BD7D','#FFCC00','#860000',"white"),c("zone5", "zone4", "zone3","zone2",  "classI", "unknown"))
osn5
osn1<-plot_individual("Z1","OSN","Olfr expression in zone1 OSN cells","individual zone1 OSN cells",c('#0505B7','#077FE2','#FFCC00','#EC001D','#860000',"white"),c("zone5","zone4","zone2", "zone1","classI", "unknown"))
osn1
osn5_drop<-plot_individual_NOunknown("Z5","OSN","Olfr expression in zone5 OSN cells","individual zone5 OSN cells",c('#0505B7','#077FE2','#23BD7D','#FFCC00','#860000'),c("zone5", "zone4", "zone3","zone2",  "classI"))
osn1_drop<-plot_individual_NOunknown("Z1","OSN","Olfr expression in zone1 OSN cells","individual zone1 OSN cells",c('#0505B7','#077FE2','#FFCC00','#EC001D','#860000'),c("zone5","zone4","zone2", "zone1","classI"))

grid<-plot_grid(inp5,inp1,iosn5,iosn1,osn5,osn1,nrow=3,ncol=2)
grid
#save_plot("Barplots_individually_zone1vzone5_dim11_mito5_noDrop.pdf",grid,base_height = 12,base_width = 10)

grid<-plot_grid(inp5_drop,inp1_drop,iosn5_drop,iosn1_drop,osn5_drop,osn1_drop,nrow=3,ncol=2)
grid
#save_plot("Barplots_individually_zone1vzone5_dim11_mito5_DropUnknown.pdf",grid,base_height = 12,base_width = 10)

###################
### SAVE LEGEND ###
###################
get_legend<-function(myggplot){
  tmp <- ggplot_gtable(ggplot_build(myggplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)
}

plot_individual_forLegend<-function(p_zone,p_cell_type,p_title,p_xlab,p_palette,p_labels){
  ggplot(data = filter(scRNA.pass3UMI.zone,zone == p_zone & cell_type == p_cell_type ),aes(x = reorder(cell_id, -total_norm_count_3UMI), y = cutoff_3umi , fill=factor(ORzone ,levels=c("5","4","3","2","1","classI") ))) +
    geom_bar(position = 'stack', stat="identity", col= "black", size=0.3) + ylim(c(0,25))+
    theme(axis.title.x=element_text(color="black",size=12,family = "Palatino"),axis.text.x=element_blank(),axis.text.y=element_text(size = 9, color="black",family = "Palatino"), 
          axis.title.y = element_text(size=12,family = "Palatino", color="black"),plot.title = element_text(hjust = 0.5,family = "Palatino",size=14), 
          legend.text=element_text(family = "Palatino", size=12), legend.title=element_text(size=14,family = "Palatino"))+
    ggtitle(p_title)+scale_x_discrete(drop=FALSE)+xlab(p_xlab)+
    scale_fill_manual(values=p_palette, labels = p_labels)+labs(fill= "OR zone") +ylab("normalized OR expression")
}

legend<-get_legend(plot_individual_forLegend("Z5","INP3","Olfr expression in zone5 INP3 cells","individual zone5 INP3 cells",zonal_fine_colors_fish_new3_inv_white,c("zone5", "zone4", "zone3","zone2", "zone1", "classI", "unknown")))
grid_legend_noDrop<-plot_grid(legend,nrow=1,ncol=1)
save_plot("legend_noDrop.pdf",grid_legend_noDrop,base_height = 2,base_width = 1)

plot_individual_NOunknown_forLegend<-function(p_zone,p_cell_type,p_title,p_xlab,p_palette,p_labels){
  ggplot(data = drop_na(filter(scRNA.pass3UMI.zone,zone == p_zone & cell_type == p_cell_type )),aes(x = reorder(cell_id, -total_norm_count_3UMI), y = cutoff_3umi , fill=factor(ORzone ,levels=c("5","4","3","2","1","classI") ))) +
    geom_bar(position = 'stack', stat="identity", col= "black", size=0.3) + ylim(c(0,25))+
    theme(axis.title.x=element_text(color="black",size=12,family = "Palatino"),axis.text.x=element_blank(),axis.text.y=element_text(size = 9, color="black",family = "Palatino"), 
          axis.title.y = element_text(size=12,family = "Palatino", color="black"),plot.title = element_text(hjust = 0.5,family = "Palatino",size=14), 
          legend.text=element_text(family = "Palatino", size=12), legend.title=element_text(size=14,family = "Palatino"))+
    ggtitle(p_title)+scale_x_discrete(drop=FALSE)+xlab(p_xlab)+
    scale_fill_manual(values=p_palette, labels = p_labels)+labs(fill= "OR zone") +ylab("normalized OR expression")
}

legend<-get_legend(plot_individual_NOunknown_forLegend("Z5","INP3","Olfr expression in zone5 INP3 cells","individual zone5 INP3 cells",zonal_fine_colors_fish_new3_inv,c("zone5", "zone4", "zone3","zone2", "zone1", "classI")))
grid_legend_Drop<-plot_grid(legend,nrow=1,ncol=1)
save_plot("legend_Drop.pdf",grid_legend_Drop,base_height = 2,base_width = 1)

###############################################################
# determine max zone to make color coded tSNE map (Figure 1C-D)
###############################################################

max_zone_meta<-scRNA.pass3UMI.zone %>% 
  filter(gene_name != is.na(gene_name)) %>% mutate(ORzone=as.character(ORzone)) %>% 
  mutate(ORzone = replace_na(ORzone, "unknown"))%>% group_by(cell_id,ORzone) %>% 
  summarize(total_cutoff_3umi=sum(cutoff_3umi)) %>% 
  group_by(cell_id) %>% top_n(1,total_cutoff_3umi) %>% rename('max_ORzone'= ORzone) %>% 
  right_join(rownames_to_column(scRNA.both@meta.data,var = 'cell_id'))%>% 
  replace_na(list(total_cutoff_3umi=0,max_ORzone='none')) %>% select(cell_id,max_ORzone,total_cutoff_3umi,orig.ident,zone) %>% distinct() %>%
  select(cell_id,max_ORzone,total_cutoff_3umi) %>%
  column_to_rownames(var = 'cell_id')

scRNA.both.working<-scRNA.both

scRNA.both.working <- AddMetaData(object = scRNA.both.working, metadata = max_zone_meta, col.name = c("max_ORzone","total_cutoff_3umi"))
TSNEPlot(scRNA.both.working, label = T)
TSNEPlot(scRNA.both.working, group.by = 'orig.ident')
TSNEPlot(scRNA.both.working, group.by = 'sort_type')
TSNEPlot(scRNA.both.working, group.by = 'max_ORzone')

p.umap.max5<-DimPlot(subset(x = scRNA.both.working, subset = zone == 'Z5'),  group.by = 'max_ORzone',reduction = "umap", label = F,pt.size = 1.5)+
  theme(plot.title = element_text(hjust = 0.5))+ 
  scale_color_manual(values = c('#EC001D','#FFCC00','#23BD7D','#077FE2','#0505B7','#860000',"gray","gray"))
p.umap.max5

p.umap.max1<-DimPlot(subset(x = scRNA.both.working, subset = zone == 'Z1'),  group.by = 'max_ORzone',reduction = "umap", label = F,pt.size = 1.5)+
  theme(plot.title = element_text(hjust = 0.5))+ 
  scale_color_manual(values = c('#EC001D','#FFCC00','#23BD7D','#0505B7','#860000',"gray","gray"))
p.umap.max1

p.umap.cluster<-DimPlot(scRNA.both.working, reduction = "umap", label = F,pt.size = 1.5)
p.umap.cluster
grid<-plot_grid(p.umap.cluster,p.umap.max1,p.umap.max5, ncol=3)
grid
#ggsave(file="UMAP.maxOR_dim11_mito5_final.pdf",grid, width = 15, height = 5, units = "in" )

#tsne
p.tsne.max5<-TSNEPlot(subset(x = scRNA.both.working, subset = zone == 'Z5'),   group.by = 'max_ORzone', label = F,pt.size = 1.5)+
  theme(plot.title = element_text(hjust = 0.5), legend.position = 'none')+ 
  scale_color_manual(values = c('#EC001D','#FFCC00','#23BD7D','#077FE2','#0505B7','#860000',"gray","gray"))
p.tsne.max5

p.tsne.max1<-TSNEPlot(subset(x = scRNA.both.working, subset = zone == 'Z1'),  group.by = 'max_ORzone', label = F,pt.size = 1.5)+
  theme(plot.title = element_text(hjust = 0.5), legend.position = 'none')+ 
  scale_color_manual(values = c('#EC001D','#FFCC00','#23BD7D','#0505B7','#860000',"gray","gray"))
p.tsne.max1

p.tsne.cluster<-TSNEPlot(scRNA.both.working, label = F,pt.size = 1.5)+theme(legend.position = "none")+
  scale_color_manual(values = c("#9999FF","#CC99FF","#9900CC","#660099","#330066","#000033"))
p.tsne.cluster

grid<-plot_grid(p.tsne.cluster,p.tsne.max1,p.tsne.max5, ncol=3)
grid
#ggsave(file="TSNE.maxOR_dim11_mito5_final.pdf",grid, width = 15, height = 5, units = "in" )

grid<-plot_grid(p.tsne.max1,p.tsne.max5, ncol=2)
grid
#ggsave(file="TSNE.maxOR_dim11_mito5_final_3.pdf",grid, width = 8, height = 4, units = "in" , useDingbats=FALSE)

p.tsne.max5_2<-TSNEPlot(subset(x = scRNA.both.working, subset = zone == 'Z5'),   group.by = 'max_ORzone', label = F,pt.size = 1.5)+
  theme(plot.title = element_text(hjust = 0.5))+ 
  scale_color_manual(values = c('#EC001D','#FFCC00','#23BD7D','#077FE2','#0505B7','#860000',"gray","gray"))
p.tsne.max5_2

grid<-plot_grid(p.tsne.max5_2, ncol=1)
grid
#ggsave(file="TSNE.maxOR_z5_dim11_mito5_final_2.pdf",grid, width = 4, height = 4, units = "in" , useDingbats=FALSE)

# TSNE PLOT OF CLUSTERING (Supp Figure 1B)

grid<-plot_grid(p.tsne.cluster,ncol=1)
grid
#ggsave(file="TSNE.cluster_2.pdf",grid, width = 4, height = 4, units = "in" ,useDingbats=FALSE)

p.tsne.sort<-TSNEPlot(scRNA.both.working,group.by = 'sort_type',label = F,pt.size = 1.5)+
  theme(legend.position = "none")+
  scale_color_manual(values = c("firebrick2","gold2","green3","darkgreen"))
p.tsne.sort

grid<-plot_grid(p.tsne.sort,ncol=1)
grid
#ggsave(file="TSNE.sorted_dim11_mito5_final.pdf",grid, width = 4, height = 4, units = "in",useDingbats=FALSE )




